8243208: Clean up JVMFlag implementation

iklam opened the pull request:
The initial commit [d85cbd6](https://github.com/openjdk/jdk/commit/d85cbd638f5b87c80eee9a92d10e1ac19d9bb2f8) is the same patch as [8243208-cleanup-jvmflag-impl.v02](https://cr.openjdk.java.net/~iklam/jdk16/8243208-cleanup-jvmflag-impl.v02/) published in [hotspot-dev@openjdk.java.net](https://mail.openjdk.java.net/pipermail/hotspot-dev/2020-September/042863.html).

The rest of the review will continue on GitHub. I will add new commits to respond to comments to the above e-mail.
<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8243208](https://bugs.openjdk.java.net/browse/JDK-8243208): Clean up JVMFlag implementation


### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 1b63ac6fc777651e26bacf22c9d9460d7e01627f
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to bf90a3765720267a8770f7bb0cc6957471ccb21b
 * [Gerard Ziemski](https://openjdk.java.net/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to 1b63ac6fc777651e26bacf22c9d9460d7e01627f


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/82/head:pull/82`
`$ git checkout pull/82`

dholmes-ora commented on src/hotspot/share/runtime/flags/jvmFlagConstraintList.cpp:
```diff
@@ -299,10 +217,11 @@ bool JVMFlagConstraintList::check_constraints(JVMFlagConstraint::ConstraintType
-    JVMFlagConstraint* constraint = at(i);
-    if (type != constraint->type()) continue;
-    if (constraint->apply(true) != JVMFlag::SUCCESS) status = false;
+  for (int i=0; i<NUM_JVMFlagsEnum; i++) {
```
Spaces around binary operators please.

iklam replied:
Fixed.

dholmes-ora commented on src/hotspot/share/runtime/flags/jvmFlagLimit.cpp:
```diff
@@ -0,0 +1,162 @@
+
+void* JVMFlagLimit::constraint_func() const {
+  int i = _constraint_func;
+  assert(0 <= i && i <NUM_JVMFlagConstraintsEnum, "sanity");
```
Space after < please

iklam replied:
Fixed.

dholmes-ora commented on src/hotspot/share/runtime/flags/jvmFlagLimit.hpp:
```diff
@@ -0,0 +1,138 @@
+    }
+  }
+
+#define AS_TYPED_LIMIT(type) inline JVMTypedFlagLimit<type>* as_ ## type() const { return (JVMTypedFlagLimit<type>*)this; }
```
Is it possible to use static_cast here rather than C-style cast?

iklam replied:
I tried but got this error:

```
invalid static_cast from type 'const JVMFlagLimit*' to type 'const JVMTypedFlagLimit<unsigned int>*'
```

dholmes-ora reviewed (approved):
One query below and a couple of style nits. Otherwise LGTM!

iklam commented:
Re: [Gerard Ziemski's comments on hotspot-dev@openjdk.java.net](https://mail.openjdk.java.net/pipermail/hotspot-dev/2020-September/042868.html):

> (1) The API JVMFlagLimit::get_constraint(flag) says “constraint”, but it returns JVMFlagLimit* using different names for the same thing - one calls it “limit", the other uses “constraint”. 

JVMFlagLimit is used for both constraints and ranges. I updated the comments:
```
// A JVMFlagLimit is created for each JVMFlag that has a range() and/or constraint() in its declaration in
// the globals_xxx.hpp file.
//
// To query the range information of a JVMFlag:
//     JVMFlagLimit::get_range(JVMFlag*)
//     JVMFlagLimit::get_range_at(int flag_enum)
// If the given flag doesn't have a range, NULL is returned.
//
// To query the constraint information of a JVMFlag:
//     JVMFlagLimit::get_constraint(JVMFlag*)
//     JVMFlagLimit::get_constraint_at(int flag_enum)
// If the given flag doesn't have a constraint, NULL is returned.
```

> (3) Isn’t there a better way to treat the special cases in "JVMFlagRangeList::print”, without explicitly checking for hardcoded function pointers? Or is it just not worth the effort, with only 2 such cases for now?

So far there are only 2 cases so I think the explicit checking may be enough. If we have more cases in the future we can have a more general solution.

> (5) I'm not sure how others feel about it, but I think I personally would find it helpful for macros like “DEFINE_RANGE_CHECK” to see a comment with a concrete example of macro expansion for some flag, just to see what it works out to be.

I think this macro is fairly easy to read, so adding an expanded example may not have much benefit:

```
#define DEFINE_RANGE_CHECK(T)                                                            \
JVMFlag::Error JVMFlagRangeChecker::check_ ## T(T value, bool verbose) const {           \
  assert(exists(), "must be");                                                           \
  JVMFlagRange_ ## T range(_flag, _limit->as_ ## T()->min(), _limit->as_ ## T()->max()); \
  return range.check_ ## T(value, verbose);                                              \
}
```

In [JDK-8081833](https://bugs.openjdk.java.net/browse/JDK-8081833), I plan to use templates to replaces the nearly identical code in jvmFlagRangeList.cpp/jvmFlagConstraintList.cpp, and these macros should go away.

(I fixed other minor nits found by Gerard.)

dholmes-ora commented on src/hotspot/share/runtime/flags/jvmFlag.cpp:
```diff
@@ -266,7 +263,7 @@ JVMFlag::MsgType JVMFlag::get_locked_message(char* buf, int buflen) const {
 
 bool JVMFlag::is_writeable() const {
-  return is_manageable() || (is_product() && is_read_write());
+  return is_manageable();
```
Unclear why is_writeable() reduces to is_manageable(). Aren't all product flags writeable? And aren't all develop flags writeable in non-product builds?

coleenp replied:
All flags shouldn't be also writeable.  Interpreter code generation depends on static value of some flags, for instance.

iklam replied:
is_writeable() doesn't mean "modifiable". The meaning is

```
// manageable flags are writeable external product flags.
//    They are dynamically writeable through the JDK management interface
//    (com.sun.management.HotSpotDiagnosticMXBean API) and also through JConsole.
```

See also [writeableFlags.cpp](https://github.com/openjdk/jdk/blob/433394203dce55db95caefcb57bac9ec114ebc47/src/hotspot/share/services/writeableFlags.cpp#L283).

There used to be `product_rw` flags, which were "writeable" but not "manageable". However, there aren't any of them anymore. So I have deleted product_rw() from the comments. 

```
/**** removed comments ****/
// product_rw flags are writeable internal product flags.
//    They are like "manageable" flags but for internal/private use.
//    The list of product_rw flags are internal/private flags which
//    may be changed/removed in a future release.  It can be set
//    through the management interface to get/set value
//    when the name of flag is supplied.
//
//    A flag can be made as "product_rw" only if
//    - the VM implementation supports dynamic setting of the flag.
//      This implies that the VM must *always* query the flag variable
//      and not reuse state related to the flag state at any given time.
//
// Note that when there is a need to support develop flags to be writeable,
// it can be done in the same way as product_rw.
```

I also removed the implementation of product_rw(). Line 269 is part of the product_rw() removal.

I considered adding support for product_rw() in the new flags implementation. It would be fairly trivial (similar to how MANAGEABLE is supported) but I am hesitant to do it because:

- No one uses it currently
- There are no test cases

I don't want to have an untested implementation which someone might accidently use in the future without any testing. If anyone indeed finds producy_rw useful, they can easily implement the functionality and can write tests for it at the same time.

dholmes-ora reviewed (approved):
Updates seem okay. One query below.

coleenp commented on src/hotspot/share/runtime/flags/jvmFlag.hpp:
```diff
@@ -34,7 +34,7 @@ class outputStream;
 typedef const char* (*RangeStrFunc)(void);
 
 struct JVMFlag {
```
I'm still getting used to git diffs, but can you change this from struct to class?  Because it looks like it does have private and public sections.

iklam replied:
I want to limit the changes in this PR, so I plan to do this in a future RFE, probably [JDK-8081833](https://bugs.openjdk.java.net/browse/JDK-8081833). I'll also change all the fields to private, and add accessor methods to them. And also add proper iterators for the flags (instead of walking JVMFlag::flags[] directly).

coleenp replied:
Ok that would be fine as a further change.

coleenp commented on src/hotspot/share/runtime/flags/jvmFlagLimit.cpp:
```diff
@@ -0,0 +1,162 @@
+  }
+  static constexpr const JVMFlagLimit* get_limit(const JVMTypedFlagLimit<T>* p, int dummy, ConstraintMarker dummy2, short func, int phase, T min, T max) {
+    return p;
+  }
```
Instead of the 'no_limit' functions, could you have get_limit have PRODUCT_RETURN(NULL) ?  Otherwise, fine to get the macros working.

iklam replied:
I think your suggestion would simply the macro materialization from
```
#ifdef PRODUCT
  ALL_FLAGS(FLAG_LIMIT_PTR_NONE,
            FLAG_LIMIT_PTR_NONE,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR_NONE,
            APPLY_FLAG_RANGE,
            APPLY_FLAG_CONSTRAINT)
#else
  ALL_FLAGS(FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR,
            APPLY_FLAG_RANGE,
            APPLY_FLAG_CONSTRAINT)
#endif
```
to 
```
ALL_FLAGS(FLAG_LIMIT_PTR_NOT_PRODUCT,
            FLAG_LIMIT_PTR_NOT_PRODUCT,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR_NOT_PRODUCT,
            APPLY_FLAG_RANGE,
            APPLY_FLAG_CONSTRAINT)
```
But I would need twice the number of the get_limit functions, like:
```
  static constexpr const JVMFlagLimit* get_limit(const JVMTypedFlagLimit<T>* p, int dummy, T min, T max) {
    return p;    // return product flag for all builds
  }
  static constexpr const JVMFlagLimit* get_limit_not_product(const JVMTypedFlagLimit<T>* p, int dummy, T min, T max) {
    NOT_PRODUCT(return p);     // return non-product flag only in non-product builds
    PRODUCT_ONLY(return NULL);  // .... but not in product build
  }
```
So the code will be more verbose than before.

Instead, in [bf90a37](https://github.com/openjdk/jdk/pull/82/commits/bf90a3765720267a8770f7bb0cc6957471ccb21b), I changed no_limit to a varargs function so now there's only one version of it.

coleenp commented on src/hotspot/share/runtime/flags/jvmFlagLookup.hpp:
```diff
@@ -0,0 +1,62 @@
+      s++;
+    }
+    return h;
+  }
```
Can this function just call hash_code(s, strlen(s)) ?

iklam replied:
strlen() is not constexpr so it can't be used here. I added my hand-rolled string_len function in [bf90a37](https://github.com/openjdk/jdk/pull/82/commits/bf90a3765720267a8770f7bb0cc6957471ccb21b) so we can share a single copy of the hash calculation loop.

coleenp replied:
Nice!

gerard-ziemski replied:
I had the same question, cool.

coleenp commented:
Looks good.  I had a couple of minor comments.
> In JDK-8081833, I plan to use templates to replaces the nearly identical code in
> jvmFlagRangeList.cpp/jvmFlagConstraintList.cpp, and these macros should go away.
Good!

iklam commented:
Re: [Gerard Ziemski's comments on hotspot-dev@openjdk.java.net](https://mail.openjdk.java.net/pipermail/hotspot-dev/2020-September/043019.html):

I made the changes in [15547eb](https://github.com/openjdk/jdk/pull/82/commits/15547ebe703fcc3a4ce154f2a01f45b756051c26)

```
Gerard #6 Do we need to provide base of the enums 
(https://hg.openjdk.java.net/jdk/jdk/raw-file/tip/doc/hotspot-style.html#enum) 
in files:

a) open/src/hotspot/share/runtime/globals_extension.hpp
   43 typedef enum {
   44   ALL_FLAGS(FLAG_MEMBER_ENUM_DEVELOP,
   45             FLAG_MEMBER_ENUM_PD_DEVELOP,
   46             FLAG_MEMBER_ENUM_PRODUCT,
   47             FLAG_MEMBER_ENUM_PD_PRODUCT,
   48             FLAG_MEMBER_ENUM_NOTPRODUCT,
   49             IGNORE_RANGE,
   50             IGNORE_CONSTRAINT)
   51   NUM_JVMFlagsEnum
   52 } JVMFlagsEnum;

b) src/hotspot/share/runtime/flags/jvmFlagLimit.cpp
   46 enum JVMFlagConstraintsEnum {
   47   ALL_CONSTRAINTS(CONSTRAINT_ENUM_)
   48   NUM_JVMFlagConstraintsEnum
   49 };
```

I added `enum ... : int` as the base.

```
Gerard #7 Should these enums be replaced with ordinary constants 
(https://hg.openjdk.java.net/jdk/jdk/raw-file/tip/doc/hotspot-style.html#enum) 
in files:

a) open/src/hotspot/share/runtime/flags/jvmFlagLimit.hpp
   44 class JVMFlagLimit {
   45   enum {
   46     HAS_RANGE = 1,
   47     HAS_CONSTRAINT = 2
   48   };
``` 

Changed to static constexpr int HAS_RANGE = 1, etc.

```
Gerard #8 Can we add a description to 
"src/hotspot/share/runtime/flags/jvmFlagLookup.hpp” on how “constexpr" 
makes this hash table possible at the compile time?
```

I added 
```
// This is a hashtable that maps from (const char*) to (JVMFlag*) to speed up
// the processing of JVM command-line argument at runtime.
//
// With constexpr, this table is generated at C++ compile time so there's
// no set up cost at runtime.
class JVMFlagLookup { ...
```

```
Gerard #9 Shouldn’t the code in JVMFlagLookup::JVMFlagLookup() initializing the 
internal tables be more like:

constexpr JVMFlagLookup::JVMFlagLookup() : _buckets(), _table(), _hashes() {
   for (int i = 0; i < NUM_BUCKETS; i++) {
     _buckets[i] = -1;
   }
   for (int i = 0; i < NUM_JVMFlagsEnum; i++) {
     _hashes[i] = 0;
     _table[i] = -1;
   }
```
There's no need to initialize _hashes and _table in a loop. Each of the elements in these two arrays are individually initialized by the expansion of this macro:
```
#define DO_HASH(flag_enum, flag_name) {          \
  unsigned int hash = hash_code(flag_name);      \
  int bucket_index = (int)(hash % NUM_BUCKETS);  \
  _hashes[flag_enum] = (u2)(hash);               \
  _table[flag_enum] = _buckets[bucket_index];    \
  _buckets[bucket_index] = (short)flag_enum;     \
}
...
  ALL_FLAGS(DO_FLAG,
            DO_FLAG,
            DO_FLAG,
            DO_FLAG,
            DO_FLAG,
            IGNORE_RANGE,
            IGNORE_CONSTRAINT)
```
There was a bug in my previous version that unnecessarily initialized ` _hashes[i] = 0` in the loop. I've removed it.

```
Gerard #10 “flagTable_verify_constexpr" and “flagTable”, in 
"src/hotspot/share/runtime/flags/jvmFlag.cpp”, are almost identical - 
can we find a way to create just one table (of the “constexpr” type)?
```

flagTable_verify_constexpr[] is needed for build-time checking only, and cannot be combined with flagTable[]. I've updated the comments to clarify:

```
// We want flagTable[] to be completely initialized at C++ compilation time, which requires
// that all arguments passed to JVMFlag() constructors to be constexpr. The following line
// checks for this this -- if any non-constexpr arguments are passed, the C++ compiler will
// generate an error.
//
// constexpr implies internal linkage. This means the flagTable_verify_constexpr[] variable
// will not be included in jvmFlag.o, so there's no footprint cost for having this variable.
//
// Note that we cannot declare flagTable[] as constexpr because JVMFlag::_flags is modified
// at runtime.
constexpr JVMFlag flagTable_verify_constexpr[] = { MATERIALIZE_ALL_FLAGS };
```

iklam commented:
/integrate

iklam commented:
Maybe it's complaining about this "*" in jvmFlagLookup.hpp? How about this:

```
  static constexpr unsigned int hash_code(const char* s, size_t len) {
    unsigned int h = 0;
    while (len -- > 0) {
-     h = 31*h + (unsigned int) *s;
+     h = ((unsigned int)31)*h + (unsigned int) *s;
      s++;
    }
    return h;
```

shipilev commented:
Seeing Windows build failures on some of my Windows build nodes as well. Filed https://bugs.openjdk.java.net/browse/JDK-8253089, please discuss there.
